pkg/storegateway/indexheader/reader_pool.go

```diff
@@ -149,7 +189,13 @@ func (p *ReaderPool) NewBinaryReader(ctx context.Context, logger log.Logger, bkt
 	}
 
 	if p.lazyReaderEnabled {
-		reader, err = NewLazyBinaryReader(ctx, readerFactory, logger, bkt, dir, id, p.metrics.lazyReader, p.onLazyReaderClosed)
+		lazyBinaryReader, lazyErr := NewLazyBinaryReader(ctx, readerFactory, logger, bkt, dir, id, p.metrics.lazyReader, p.onLazyReaderClosed)
```

Comment:
The flow for eager loading a bit hard to follow. It's code that's executed on every block load as opposed to only on startup. One contrived example is when a block changes owners and a block will be eagerly loaded instead of lazy loaded:

* st-gw-0 owns the block and has it loaded
* st-gw-0 restarts and eagerly loads the block
* st-gw-1 comes up and starts owning the block
* st-gw-1 becomes unhealthy or is scaled down or changes tokens
* st-gw-0 starts owning the block again. Instead of having it as a lazy-loaded block, st-gw-0 immediately eagerly loads it because it's in the snapshot

I wonder if we shouldn't remove a block from the snapshot once we've loaded it once already.

Reply:
After I look again, indeed this is wrong. Currently, this will eager load the header on every syncBlock ticker. 

I pushed a modification of adding one flag `initialSync` to `NewBinaryReader` contructor. The flag will help to decide whether lazyBinaryReader will do eager loading or not. And we are setting up in such a way eagerLoading when enabled only be done on initialSync block.

```
        if p.lazyReaderEnabled {
		lazyBinaryReader, lazyErr := NewLazyBinaryReader(ctx, readerFactory, logger, bkt, dir, id, p.metrics.lazyReader, p.onLazyReaderClosed)
		// we only try to eager load only during initialSync
		if initialSync && p.eagerLoadReaderEnabled && p.preShutdownLoadedBlocks != nil {
			// we only load if lazyBinaryReader is non nil and we have preShutdownLoadedBlocks
			if lazyErr == nil && p.preShutdownLoadedBlocks.IndexHeaderLastUsedTime[id] > 0 {
				lazyBinaryReader.EagerLoad()
			}
		}
		reader, err = lazyBinaryReader, lazyErr
	} else {
		reader, err = readerFactory()
	}
```

I will check again to see if this work and making sure nothing breeak.

Please help to take a look and let me know what you think.